qiskit_experiments/framework/experiment_data.py

```diff
@@ -767,6 +770,66 @@ def add_data(
     ) -> None:
         """Add experiment data.
 
+        Args:
+            data: Experiment data to add. Several types are accepted for convenience:
+
+                * Result: Add data from this ``Result`` object.
+                * List[Result]: Add data from the ``Result`` objects.
+                * Dict: Add this data.
+                * List[Dict]: Add this list of data.
+
+        Raises:
+            TypeError: If the input data type is invalid.
+        """
+
+        if any(not future.done() for future in self._analysis_futures.values()):
+            LOG.warning(
+                "Not all analysis has finished running. Adding new data may "
+                "create unexpected analysis results."
+            )
+        if not isinstance(data, list):
+            data = [data]
+
+        # Directly add non-job data
+        with self._result_data.lock:
+            tmp_exp_data = ExperimentData()
+            composite_flag = False
+            experiment_seperator = {}
+            for datum in data:
+                if isinstance(datum, dict):
+                    if "composite_metadata" in datum["metadata"]:
+                        composite_flag = True
+                        marginalized_data = self._marginalized_component_data([datum])
+                        for inner_datum in marginalized_data:
+                            #print(inner_datum)
+                            if "experiment_type" in inner_datum[0]["metadata"]:
```

Comment:
You need to consider the case that composite experiment is nested. This experiment_type can be some composite experiment subclass. If you encounter this situation, you need to create another layer in the experiment data tree. I suggest turning the code inside this loop into a separate protected method (or inner function) so that you can call it recursively.